shell/src/main/java/alluxio/cli/fs/command/PersistCommand.java

```diff
@@ -55,63 +85,103 @@ public void validateArgs(CommandLine cl) throws InvalidArgumentException {
   @Override
   protected void runPlainPath(AlluxioURI plainPath, CommandLine cl)
       throws AlluxioException, IOException {
-    persist(plainPath);
+    URIStatus status = mFileSystem.getStatus(plainPath);
+    queuePersist(status);
+  }
+
+  private void queuePersist(URIStatus status) throws AlluxioException, IOException {
+    AlluxioURI uri = new AlluxioURI(status.getPath());
+    if (status.isFolder()) {
+      List<URIStatus> statuses = mFileSystem.listStatus(uri);
+      for (URIStatus s : statuses) {
+        queuePersist(s);
+      }
+    } else if (status.isPersisted()) {
+      System.out.println(status.getPath() + " is already persisted, skipping.");
+    } else {
+      mToPersist.add(uri);
+    }
   }
 
   @Override
   public int run(CommandLine cl) throws AlluxioException, IOException {
+    resetState();
+    int parallelism = 1;
+    if (cl.hasOption(PARALLELISM_OPTION.getLongOpt())) {
+      String parellismOption = cl.getOptionValue(PARALLELISM_OPTION.getLongOpt());
+      parallelism = Integer.parseInt(parellismOption);
+    }
     String[] args = cl.getArgs();
     for (String path : args) {
       AlluxioURI inputPath = new AlluxioURI(path);
       runWildCardCmd(inputPath, cl);
     }
-    return 0;
-  }
-
-  /**
-   * Persists a file or directory currently stored only in Alluxio to the UnderFileSystem.
-   *
-   * @param filePath the {@link AlluxioURI} path to persist to the UnderFileSystem
-   */
-  private void persist(AlluxioURI filePath) throws AlluxioException, IOException {
-    URIStatus status = mFileSystem.getStatus(filePath);
-    if (status.isFolder()) {
-      List<URIStatus> statuses = mFileSystem.listStatus(filePath);
-      List<String> errorMessages = new ArrayList<>();
-      for (URIStatus uriStatus : statuses) {
-        AlluxioURI newPath = new AlluxioURI(uriStatus.getPath());
-        try {
-          persist(newPath);
-        } catch (Exception e) {
-          errorMessages.add(e.getMessage());
-        }
-      }
-      if (errorMessages.size() != 0) {
-        throw new IOException(Joiner.on('\n').join(errorMessages));
-      }
-    } else if (status.isPersisted()) {
-      System.out.println(filePath + " is already persisted");
-    } else {
-      try {
-        FileSystemUtils.persistFile(mFileSystem, filePath);
-      } catch (InterruptedException e) {
-        Thread.currentThread().interrupt();
-        throw new RuntimeException(e);
-      } catch (TimeoutException e) {
-        throw new RuntimeException(e);
+    mTotalFilesToPersist = mToPersist.size();
+    System.out.println("Found " + mTotalFilesToPersist + " files to persist.");
+    ExecutorService service = Executors.newFixedThreadPool(parallelism);
```

Comment:
should we make these daemon threads?

Reply:
What are the advantages? These don't feel like daemon threads to me, since we are essentially joining on them from the main thread.

Reply:
If the main thread is interrupted and finishes, should the JVM stick around while the persist threads complete?

Reply:
Are you thinking about this from a testing perspective (to prevent the JVM from hanging in some weird cases we are testing)? It doesn't seem reasonable for the main thread to be interrupted without also interrupting these threads.
There definitely is a risk of these threads running for a long time, since each persist call blocks up to 20 minutes, I hoped that the interrupt handling would be sufficient to address this kind of problem.
On the other hand, I don't think there is much benefit (aside from making the code read better) of making these nondaemonic because the main thread joins on them. Do you have a strong opinion here?

Reply:
I've added thread clean up logic and made them daemon threads.